### PR TITLE
Refinement near a boundary

### LINKED ISSUE #17661: BoundaryMarker
## Reason
For initial adaptivity it would be convenient to be able to mark all elements adjacent to a given boundary for refinement (or coarsening).

## Design
Use boundary info data or stuff from MooseMesh.

## Impact
Added capability.

### PR DESCRIPTION
**Pull request by dschwen**
Closes #17661

### CODE REVIEW

File: framework/doc/content/source/markers/BoundaryMarker.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+!syntax description /Adaptivity/Markers/BoundaryMarker
+
+## Description
+
+The `BoundaryMarker` is a stand-alone marker that marks all
+elements either directly or within a given `distance` to a boundary.
+
+!alert warning title=Distributed Mesh
+Currently this Marker may give inaccurate results when using distributed meshes
+while also supplying the `distance` parameter. Parts of the boundary on different
```
Comment (loganharbour) [inline framework/doc/content/source/markers/BoundaryMarker.md:12]:
```
[!param](/Adaptivity/Markers/BoundaryMarker/distance)
```

for all listings of distance

File: framework/doc/content/source/markers/BoundaryMarker.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
+# BoundaryMarker
+
+!syntax description /Adaptivity/Markers/BoundaryMarker
+
+## Description
+
+The `BoundaryMarker` is a stand-alone marker that marks all
+elements either directly or within a given `distance` to a boundary.
+
+!alert warning title=Distributed Mesh
```
Comment (loganharbour) [inline framework/doc/content/source/markers/BoundaryMarker.md:10]: Should you not put a mooseWarning in the code for this? Non-deterministic behavior in parallel is not so good. That or gather the points for distributed mesh.

File: framework/include/markers/BoundaryMarker.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+  BoundaryMarker(const InputParameters & parameters);
+
+protected:
+  virtual MarkerValue computeElementMarker() override;
+
+  const Real _distance;
+  const std::unordered_map<boundary_id_type, std::unordered_set<dof_id_type>> & _bnd_elem_ids;
+  MarkerValue _mark;
+  BoundaryID _boundary;
```
Comment (loganharbour) [inline framework/include/markers/BoundaryMarker.h:30]: const some more things and docstrings?

File: framework/src/markers/BoundaryMarker.C [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    return DONT_MARK;
+  }
+  else
+  {
+    const auto it = _bnd_elem_ids.find(_boundary);
+    if (it != _bnd_elem_ids.end())
+      for (const auto id : it->second)
+      {
+        const auto elem = _mesh.elemPtr(id);
+        if (elem)
```
Comment (loganharbour) [inline framework/src/markers/BoundaryMarker.C:55]: If you're checking for existence of an element here, you want queryElemPtr(). However - I don't think you should be checking for existence. If the map isn't valid, that's bad I think? You should have only elem ids in there that you know about in distributed mesh, no?

**Review by loganharbour — CHANGES_REQUESTED** [review]

File: framework/doc/content/source/markers/BoundaryMarker.md (hunk 2 above)
Comment (dschwen) [reply framework/doc/content/source/markers/BoundaryMarker.md:10]: Yeah, I guess 🙎

File: framework/src/markers/BoundaryMarker.C (hunk 4 above)
Comment (dschwen) [reply framework/src/markers/BoundaryMarker.C:55]: Ok, if I can rely on that, that's fine

**Comment by dschwen** [comment]
Another vexing crash. I've invalidated. Test passes locally and the on change to `BoxMarker` should really not affect this....

File: framework/include/markers/BoundaryMarker.h [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  const Real _distance;
+
+  /// lists of boundary elements for all boundaries
+  const std::unordered_map<boundary_id_type, std::unordered_set<dof_id_type>> & _bnd_elem_ids;
+
+  /// which way to mark elements near teh boundary
+  MarkerValue _mark;
+
+  /// boundary near which to mark elements
+  BoundaryID _boundary;
```
Comment (loganharbour) [inline framework/include/markers/BoundaryMarker.h:37]:
```suggestion
  /// which way to mark elements near the boundary
  const MarkerValue _mark;

  /// boundary near which to mark elements
  const BoundaryID _boundary;
```

**Review by loganharbour — CHANGES_REQUESTED** [review]

**Review by loganharbour — APPROVED** [review]

**Pull request merged**